config.py

```diff
@@ -34,57 +34,92 @@
 c['slavePortnum'] = 9989
 
 from flocker_bb.password import generate_password
-from flocker_bb.ec2_buildslave import EC2BuildSlave
+from flocker_bb.ec2 import rackspace_slave, ec2_slave
 from buildbot.buildslave import BuildSlave
 
-cloudInit = FilePath(__file__).sibling("slave").child("cloud-init.sh").getContent()
+cloudInit = FilePath(__file__).sibling("slave").child(
+    "cloud-init.sh").getContent()
 
 c['slaves'] = []
 SLAVENAMES = {}
 
+
+def get_cloud_init(name, base, password, provider, privateData, slavePortnum):
+    """
+    :param bytes name: The name of the buildslave.
+    :param bytes base: The name of image used for the buildslave.
+    :param bytes password: The password that the buildslave will use to
+       authenticate with the buildmaster.
+    :param bytes provider: The cloud ``provider`` hosting the buildslave. This
+        is added to an environment variable, so that cloud ``provider``
+        specific tests know which cloud authentication plugin to load and which
+        credentials to load from the credentials file.
+    :param dict privateData: The non-slave specific keys and values from the
+        buildbot ``config.yml`` file.
+    :parma int slavePortnum: The TCP port number on the buildmaster to which
+        the buildslave will connect.
+    :returns: The ``bytes`` of a ``cloud-init.sh`` script which can be supplied
+       as ``userdata`` when creating an on-demand buildslave.
+    """
+    return cloudInit % {
+        "github_token": privateData['github']['token'],
+        "coveralls_token": privateData['coveralls']['token'],
+        "name": name,
+        "base": base,
+        "password": password,
+        "FLOCKER_FUNCTIONAL_TEST_CLOUD_PROVIDER": provider,
+        'buildmaster_host': privateData['buildmaster']['host'],
+        'buildmaster_port': slavePortnum,
+        'acceptance.yml': privateData['acceptance'].get('config', ''),
+        'acceptance-ssh-key': privateData['acceptance'].get('ssh-key', ''),
+    }
+
+
 for base, slaveConfig in privateData['slaves'].items():
     SLAVENAMES[base] = []
-    if 'ami' in slaveConfig:
+    if "openstack-image" in slaveConfig:
+        password = generate_password(32)
+        slave = rackspace_slave(
+            name=base,
+            password=password,
+            config=slaveConfig,
```

Comment:
I was kind of imagining there'd be a separate object covering the common pieces and the code could be like `ondemand_slave(cloud_specific_object, **ondemand_bits)`.